### PR TITLE
Reimplement `wasmtime-wasi` on top of `wasmtime` 

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit reimplements the `wasmtime-wasi` crate on top of the
`wasmtime` API crate, instead of being placed on top of the `wasmtime-*`
family of internal crates. The purpose here is to continue to exercise
the API as well as avoid usage of internals wherever possible and
instead use the safe API as much as possible.

The `wasmtime-wasi` crate's API has been updated as part of this PR as
well. The general outline of it is now:

* Each module snapshot has a `WasiCtxBuilder`, `WasiCtx`, and `Wasi`
  type.
  * The `WasiCtx*` types are reexported from `wasi-common`.
  * The `Wasi` type is synthesized by the `wig` crate's procedural macro
* The `Wasi` type exposes one constructor which takes a `Store` and a
  `WasiCtx`, and produces a `Wasi`
* Each `Wasi` struct fields for all the exported functions in that wasi
  module. They're all public an they all have type `wasmtime::Func`
* The `Wasi` type has a `get_export` method to fetch an struct field by
  name.

The intention here is that we can continue to make progress on #727 by
integrating WASI construction into the `Instance::new` experience, but
it requires everything to be part of the same system!

The main oddity required by the `wasmtime-wasi` crate is that it needs
access to the caller's `memory` export, if any. This is currently done
with a bit of a hack and is expected to go away once interface types are
more fully baked in.

### CODE REVIEW

File: crates/wasi/src/old/snapshot_0.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+use wasi_common::old::snapshot_0::hostcalls;
+use wasi_common::old::snapshot_0::WasiCtx;
+
+// Defines a `struct Wasi` with member fields and appropriate APIs for dealing
+// with all the various WASI exports.
+wig::define_wasi_struct!(
+    "old/snapshot_0" "wasi_unstable"
+);
+
+pub fn is_wasi_module(name: &str) -> bool {
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wasi/src/old/snapshot_0.rs]: Currently this is used during instantiation for the python and interface types extensions, but I'd like to delete those as soon as we better integrate wasi into the `wasmtime` crate. So in that sense just a temporary holdover!

File: crates/wasi-common/wig/src/wasi.rs [hunk 2]
```diff
@@ -178,56 +182,71 @@ pub fn add_wrappers_to_module(args: TokenStream) -> TokenStream {
...
-        ) {
-            #(#add)*
+        /// An instantiated instance of the wasi exports.
+        ///
+        /// This represents a wasi module which can be used to instantiate other
+        /// wasm modules. This structure exports all that various fields of the
+        /// wasi instance as fields which can be used to implement your own
+        /// instantiation logic, if necessary. Additionally [`Wasi::get_export`]
+        /// can be used to do name-based resolution.
+        pub struct Wasi {
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wasi-common/wig/src/wasi.rs]: I think the current scheme of one-struct-of-functions isn't gonna fly for very long. Once we get modularization I suspect we'll have a struct-of-structs-of-functions where we'll have a struct-per-module and one struct to bring them all together (maybe with options?). Or something like that, maybe not even public fields at all and just the lookup function!

Overall I don't have too many thoughts about how exactly this should get organized, I'm down for whatever we think feels right for wasi. I definitely think our initialization story will need a bit more evolution before we expose this from the `wasmtime` crate itself (or some equivalent thereof)

File: crates/wasi/src/old/snapshot_0.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (peterhuene) [reply crates/wasi/src/old/snapshot_0.rs]: I specifically meant this copy of `is_wasi_module`, though.  The other one in `lib.rs` is definitely used.  I couldn't find any references to this one, so I was wondering if it was accidentally duplicated when we copied the source for the last snapshot.

[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wasi/src/old/snapshot_0.rs]: Oh right! Yeah this is entirely unused, and I think it's just sticking around from the "let's copy everything" transition phase between the last two snapshots.

**Pull request merged**